ideascube/serveradmin/catalog.py

```diff
@@ -176,6 +185,97 @@ def unregister_site(self, package):
         except FileNotFoundError as e:
             printerr(e)
 
+class MediaCenter(Handler):
+    def commit(self):
+        for pkg in self._removed:
+            self.remove_pkg(pkg)
+        for pkg in self._installed:
+            self.install_pkg(pkg)
+        super().commit()
+
+    def _install_media(self, media_info, install_dir, package_id):
+        #ensure that mandatory fields are set
+        for entry in ('title', 'summary', 'path', 'credits'):
+            if entry not in media_info:
+                raise InvalidPackageContent('Missing {} in {}'.format(entry, media_info))
+
+        files = {'original': None, 'preview': None}
+
+        title = smart_truncate(media_info['title'])
+
+        lang = media_info.get('lang', settings.LANGUAGE_CODE)
+
+        kind = media_info.get('kind')
+        if not kind or not hasattr(Document, kind.upper()):
+            content_type, _ = mimetypes.guess_type(original)
+            kind = guess_kind_from_content_type(content_type) or Document.OTHER
+
+        instance = Document.objects.filter(title=title, kind=kind).last()
+        if instance:
+            raise InvalidPackageContent('Document {} already exists'.format(title))
+
+        metadata = {
+            'title'      : title,
+            'summary'    : media_info['summary'],
+            'credits'    : media_info['credits'],
+            'tags'       : media_info.get('tags'),
+            'lang'       : lang,
+            'kind'       : kind,
+            'package_id' : package_id
+        }
+
+        o, p = None, None
+        try:
+            original = media_info['path']
+            try:
+                o = Path(install_dir, original).open('rb')
+                files['original'] = File(o, name=original)
+            except OSError as e:
+                raise InvalidPackageContent('Cannot open path {}. Error is {}'.format(path, e))
+
+            if 'preview' in media_info:
+                preview = media_info['preview']
+                try:
+                    p = Path(install_dir, preview).open('rb')
+                    files['preview'] = File(p, name=preview)
+                except OSError:
+                    pass
+
+            self._save_media(metadata, files)
+        finally:
```

Comment:
The first/big try/catch replace the `with open(...) as f:` pattern.  
But I'm not fan of nested `with ...` that end to different calls to `self._save_media` depending of the context (as it is made in import_medias command)

We could somehow merge the first nested `try/except OSError` in the big one. But not sure for the second one.

Btw, we will not keep this code cause we don't what to copy the file but use the already installed one by the catalog. Not sure it worst the change here if we drop the code latter.